[xbuild] install cargo based tooling as specified versions.

rexhoffman opened the pull request:
## Motivation

The first step of xbuild is making sure sccache is installed.   So implemented tooling in x to allow automatic installation of specific versions of cargo managed tooling.   

Right not tracking sccache and grcov, but more can easily be added.   Supports automatic install with:
```cargo x tools```, and can be checked with ```cargo x tools --check```.

### Have you read the [Contributing Guidelines on pull requests]

Yes

## Test Plan

Currently tested manually, but looking at ways to simulate cargo commands, may add in the near future.

## Related PRs

None

bmwill commented on devtools/x/src/build.rs:
```diff
@@ -0,0 +1,140 @@
+    /// Run test on the provided packages
+    package: Vec<String>,
+    #[structopt(long, short)]
+    /// Skip running expensive libra testsuite integration tests
+    unit: bool,
+    #[structopt(long)]
+    /// Test only this package's library unit tests, skipping doctests
+    lib: bool,
+    #[structopt(long)]
+    /// Do not fast fail the run if tests (or test executables) fail
+    no_fail_fast: bool,
+    #[structopt(long)]
+    /// Do not run tests, only compile the test executables
+    no_run: bool,
+    #[structopt(long, short)]
+    /// Number of parallel jobs, defaults to # of CPUs
+    jobs: Option<u16>,
+    #[structopt(long, parse(from_os_str))]
+    /// Directory to output HTML coverage report to (using grcov)
+    html_cov_dir: Option<OsString>,
+    #[structopt(name = "TESTNAME", parse(from_os_str))]
+    testname: Option<OsString>,
+    #[structopt(name = "ARGS", parse(from_os_str), last = true)]
+    args: Vec<OsString>,
```
This looks like it was mostly copied from `xtest` and there's a couple of these params that maybe don't make sense wrt wrapping `cargo build`.

rexhoffman replied:
ah didn't mean to include this yet.

bmwill commented on devtools/x/src/build.rs:
```diff
@@ -0,0 +1,140 @@
+    args.args.extend(args.testname.clone());
+    let config = xctx.config();
+
+    let env_vars: &[(&str, &str)] = if args.html_cov_dir.is_some() {
```
Does `xbuild` need to know anything about code coverage output? seems like that's specific to `xtest`

rexhoffman replied:
grcov is text, yes, seemed reasonable to allow it all be installed up front to potentially replace parts of dev_setup.sh.

bmwill commented on devtools/x/src/installer.rs:
```diff
@@ -0,0 +1,38 @@
+pub fn install_if_needed_or_die_config(config: &CargoConfig, name: &str, version: &str) {
+    if !check_installed(name, version) {
+        info!("Installing {} {}", name, version);
+        let installres = Cargo::new(config, "install")
```
```suggestion
        let installers = Cargo::new(config, "install")
```

bmwill commented on devtools/x/src/installer.rs:
```diff
@@ -0,0 +1,38 @@
+use log::info;
+use std::process::Command;
+
+pub fn install_if_needed_or_die(name: &str, version: &str) {
```
Is this only supposed to install a different cargo version? (I'm assuming yes)

rexhoffman replied:
Don't want to use the nightly version that could cause breakage in these released version of tools.

rexhoffman replied:
if we don't have the current cargo stable locally, totally fine with using that to build the latest tooling, since the stable version we're using isn't in x.toml -- though weirdly the nightly version is.   This all seems to need a reworking but not one I wanted to take on now with the redundant info in x.toml and cargo-toolchain.

bmwill commented on devtools/x/src/installer.rs:
```diff
@@ -0,0 +1,38 @@
+}
+
+pub fn check_installed(name: &str, version: &str) -> bool {
+    let result = Command::new(name).arg("--version").output();
+    let comp = format!("{} {}", name, version);
+    return result.is_ok()
+        && comp.eq(String::from_utf8_lossy(&result.unwrap().stdout.as_slice()).trim());
```
This could probably be simplified a little bit using a `match` or `if let`

rexhoffman replied:
used match, doesn't feel simpler to be, but probably more canonically rust.

bmwill commented on devtools/x/src/tools.rs:
```diff
@@ -0,0 +1,36 @@
+    let mut failed: bool = false;
+    for (key, value) in tools {
+        if !args.check {
+            installer::install_if_needed_or_die(key, value);
```
Why or die? might be better to have this return an error and handle the exit code in 1 single place (like you are doing below)

rexhoffman replied:
The plan is to use the installer in multiple locations, I suppose I can try to find a good one liner to wrap the calls when used in xtext/xbuild.    The `x tools` is mostly about pre-installation in CI environments, and if a dev wants to setup for offline work.

rexhoffman replied:
removed, the or_die bit.   will need a good one line when used in xtest/xbuild.

rexhoffman commented:
@bmwill addressed your concerns, but the logging is a little less user friendly, and a little more verbose now as a bit of code golf has to be played to make the use of return values you seemed to want with matches/if lets.   Ultimately I'm okay with since the extra logging every time someone will run xtest with coverage of xbuild with sccache will just inform them of the tools they are actually using.   Not attempting to address the cargo version mismatch we are in, as @metajack seems to think it will be resolved with the next release of rust.

bmwill commented on devtools/x/src/installer.rs:
```diff
@@ -0,0 +1,53 @@
+
+pub fn check_installed(name: &str, version: &str) -> bool {
+    let result = Command::new(name).arg("--version").output();
+    let found = match result {
+        Ok(_) => format!("{} {}", name, version)
+            .eq(String::from_utf8_lossy(&result.unwrap().stdout.as_slice()).trim()),
+        Err(_) => false,
+    };
```
```suggestion
    let found = match result {
        Ok(output) => format!("{} {}", name, version)
            .eq(String::from_utf8_lossy(&output.stdout).trim()),
        Err(_) => false,
    };
```

bmwill replied:
In order to remove the need to unwrap

bmwill commented on devtools/x/src/config.rs:
```diff
@@ -152,4 +153,8 @@ impl Config {
     }
+
+    pub fn tools(&self) -> &[(String, String)] {
+        &self.tools.as_slice()
```
```suggestion
        &self.tools
```

bmwill commented on devtools/x/src/installer.rs:
```diff
@@ -0,0 +1,53 @@
+            .eq(String::from_utf8_lossy(&output.stdout).trim()),
+        Err(_) => false,
+    };
+    info!(
+        "{} of version {} is{} installed",
+        name,
+        version,
+        if !found { " not" } else { "" }
+    );
```
So I don't think we want this to print out every time anyone ever runs xbuild or xtest. So this should either be a lower level or only output in the error case

rexhoffman replied:
It wont, it'll only print out if we check for it.   Which will occur only when needed.   So I'm okay with it printing out before use.

bmwill replied:
What I'm saying is I don't want a messages saying "sccache is installed" every time I run  `cargo xbuild`, assuming xbuild uses sccache. It makes sense to print a message out the first time something is installed, but not every time the tool is checked for

rexhoffman replied:
I'm invisioning the log line stating that sccache is installed at a certain version occurring right before the env vars are set and logged, and the build line is logged.   Ideally I want folks to know exactly what's going on underneath the covers when xbuild runs, and seeing the sccache log line makes sense to me.   Young has used the log line of what's being built in the past to split out tests in to different runs....  I just want people to be able to drop to the command line and do exactly the same thing as the build, so I don't get called in to debug someone's build if I can avoid it.

bmwill reviewed (approved):
I would just fix the last nit I had before landing.

rexhoffman commented:
/land
